File: 2024/solutions/go/01/solution.go
package main

import (
	"fmt"
	"math"
	"os"
	"slices"
	"strconv"
	"strings"
)

func main() {
	text, err := os.ReadFile("../../../input/01")
	if err != nil {
		fmt.Println("Error reading input file")
	}
	lines := strings.Split(string(text), "\n")

	fmt.Println(partOne(lines))
}

func partOne(lines []string) int {
	var leftSideLocationIds = make([]int, len(lines))
	var rightSideLocationIds = make([]int, len(lines))

	for index, line := range lines {
		locationIdStrings := strings.Split(line, "   ")
		leftSideLocationId, err := strconv.Atoi(locationIdStrings[0])
		if err != nil {
			fmt.Printf("Error converting left side of line %d to integer", index)
		}
		leftSideLocationIds[index] = leftSideLocationId
		rightSideLocationId, err := strconv.Atoi(locationIdStrings[1])
		if err != nil {
			fmt.Printf("Error converting right side of line %d to integer", index)
		}
		rightSideLocationIds[index] = rightSideLocationId
	}

	slices.Sort(leftSideLocationIds)
	slices.Sort(rightSideLocationIds)

	var diffs = make([]int, len(lines))

	for index, leftSideLocationId := range leftSideLocationIds {
		rightSideLocationId := rightSideLocationIds[index]
		// will there be some loss of precision by this conversion from integers to
		// floats and then back to an integer?
		diffs[index] = int(math.Abs(float64(leftSideLocationId) - float64(rightSideLocationId)))
	}

	totalDiff := 0

	for _, diff := range diffs {
		totalDiff += diff
	}

	return totalDiff
}

File: 2024/solutions/typescript/01/solution.ts
const lines = (
	await Bun.file(`${process.env.ASSETS_REPO}/2024/input/01`).text()
).split("\n");

type LocationId = number;

function partOne() {
	const leftSideLocationIds: Array<LocationId> = [];
	const rightSideLocationIds: Array<LocationId> = [];

	for (const line of lines) {
		const [leftSideLocationId, rightSideLocationId] = line
			.split("   ")
			.map((locationIdString) => Number.parseInt(locationIdString, 10));
		leftSideLocationIds.push(leftSideLocationId);
		rightSideLocationIds.push(rightSideLocationId);
	}

	leftSideLocationIds.sort();
	rightSideLocationIds.sort();

	const diffs: Array<number> = [];

	leftSideLocationIds.forEach((leftSideLocationId, index) => {
		const rightSideLocationId = rightSideLocationIds[index];
		diffs.push(Math.abs(leftSideLocationId - rightSideLocationId));
	});

	return diffs.reduce(
		(accumulator, currentNumber) => accumulator + currentNumber,
		0,
	);
}

function partTwoSolutionOne() {
	const leftSideLocationIdMatchesOnRightSide: Record<LocationId, number> = {};
	const rightSideLocationIds: Array<LocationId> = [];

	for (const line of lines) {
		const [leftSideLocationId, rightSideLocationId] = line
			.split("   ")
			.map((locationIdString) => Number.parseInt(locationIdString, 10));
		leftSideLocationIdMatchesOnRightSide[leftSideLocationId] = 0;
		rightSideLocationIds.push(rightSideLocationId);
	}

	for (const locationId of rightSideLocationIds) {
		if (typeof leftSideLocationIdMatchesOnRightSide[locationId] === "number") {
			leftSideLocationIdMatchesOnRightSide[locationId]++;
		}
	}

	return Object.entries(leftSideLocationIdMatchesOnRightSide).reduce(
		(accumulator, currentMatch) => {
			const locationId = Number.parseInt(currentMatch[0], 10);
			const matchesOnRightSide = currentMatch[1];
			return accumulator + locationId * matchesOnRightSide;
		},
		0,
	);
}

function partTwoSolutionTwo() {
	const leftSideLocationIds: Record<LocationId, boolean> = {};
	const rightSideLocationIds: Array<LocationId> = [];

	for (const line of lines) {
		const [leftSideLocationId, rightSideLocationId] = line
			.split("   ")
			.map((locationIdString) => Number.parseInt(locationIdString, 10));
		leftSideLocationIds[leftSideLocationId] = true;
		rightSideLocationIds.push(rightSideLocationId);
	}

	let similarityScore = 0;

	for (const rightSidelocationId of rightSideLocationIds) {
		if (leftSideLocationIds[rightSidelocationId]) {
			similarityScore += rightSidelocationId;
		}
	}

	return similarityScore;
}

console.log(partOne());
console.log(partTwoSolutionOne());
console.log(partTwoSolutionTwo());

export { partOne, partTwoSolutionOne, partTwoSolutionTwo };
